### ISSUE TITLE
RFE: test metadata (L1) setup and cleanup

### ISSUE THREAD

**Issue by jscotka** [psss/tmt#825]
Hi, we've found that it would be very useful, that it would be nice to have something like `prepare` but for test metadata (L1)
so that there is `test:` and I propose also to be able to declare also `setup` and `cleanup` to run before and after test.

  - First PoC should be very easy to implement (more less same as test handling)
  - It could extend behaviour of TMT to be able to distinguish when preparation or test failed for any test type. not just beakerlib test what handles it internally. 
  - It could add conditions, if continue with execution or not. e.g. if preparation fails, if still continue with test, or condition if call cleanup if setup failed.
  - I can imagine that behaviour could be clever and e.g. in case setup is set upper in FMF so that optimise calling of setup and cleanup scripts to avoid calling it more times than expected (could be also configurable by some condition, if call it once or everytime)

**UserStory:** I want to distinguish if preparation of environment for test failed.

 - Currently It is possible to have this part as part of `test` itself, what is probably bad idea
 - Another solution is to move it to prepare of `plan` and then just not exectute test if preparation fails, but some of these preparation are more closer to tests than plans.
 - Theoretical solution is also to have plan of plans what will also solve this situation. But probably it is more powerful for this usecase. 
 - Or I can imagine also somehow connect tests together, and add there some type spefification by extension of link, e.g. `link: {setup-by: /path/to/setup/test, cleanup-by: /cleanup/test}` but probably it then needs more clever logic inside to parse and optimise test execution order and differenciate test types, but it seems like very very flexible, but harder solution, applicable.
   - It reminds me, that it could be also way how to implement plan of plans via some `link: [ {implemented-by: url//to/another/plan1}, {implemented-by: url//to/another/plan2}] ` where url could be also some url and ref to indentify also git and proper commit hash or branch or tag.

**Comment by jscotka** [comment]
few possible examples:
content of `curl.fmf`
```yaml
test: ./runtest.sh
summary: Test curl against
cleanup: ./stop_server.sh

/apache:
  summary+: apache 
  setup: ./setup_apache.sh
/ngnix:
  summary+: ngnix
  setup:
    - ./setup_ngnix.sh
    - ./setup_usera.sh
```
Pros: 
 - With same test for `curl` I can test it against apache or ngnix server very easily.
 - When `setup` of e.g. `apache` fails it does not make sense to run the test itself. because preparation is invalid
 - May report WARN or ERROR, not FAIL without beakerlib, where it is supported
 - It is simple solution, easy to implement
 - allow connect setup with test instead of preparation in plan

Cons:
 - specific just to tests.
 
It may also contain more generic than just contain command to execute.
I can imagine some syntax like:
```yaml
link:
   setup-by:
     url: url://github/user/repo
     ref: devel_branch
     [id|filter]: /some/path/regexp or some filter of the test
```

what will identify, that you have to run some test before this one and the mentioned test will work as setup test, and in case this setup test fails it does not make sense to run this test (with some good defaults, no `url` meas this repo,  no `ref` means this branch, so theoretically just test id or some filter is necessary.) one setup may request another setup (DDT idea)

Pros:
 - it is generic graph implementation (some constrains to avoid graph problems)
 - allow very flexible referencing (potentially linking remote repositories)
 - when we solve it for this for setup, it could be then used generally, theoretically, resolving tuples of (`X:` and `X-by:`)
 - and generic resolver of depepncecies between tests and plans.

Cons:
 - will need some more complex linting
 - ensure that order will be same by some rules to have repeatable output
 - need to identify type, that some test is setup

Theoretically it will solves also plan of plans. Finally it may any plan may reference another plans, so that plan may reference plan what will reference another plan.

e.g. `daemons-tier.fmf`:
```yaml
link:
   - implemented-by:
     url: git://some/git/repo/component1
     ref: some_commit
     [filter|id]: [tag:Tier1|/plans/identifier]
```

**Comment by adiosnb** [comment]
This proposal may be interesting, especially if `beakerlib`, `pytest`, or any different test framework is omitted. If a test is split into four generic steps: `setup`, `exercise`, `verify`, and `teardown`, all of them are somehow implemented in test frameworks (e.g., beakerlib). However, the is no easy possibility to define these steps generically in the tmt test without these frameworks.

The exercise and verify steps are part of a script under the `test:` keyword. This proposal would add the remaining steps (setup, teardown/cleanup) in the tmt test and allow to share and extend scripts between tmt tests.

+1 for me

`test` keyword in a tmt test is a shell command. Thus, I would also propose to have shell commands in `setup` and `cleanup` entries. Of course, the FMF ID would create a more robust solution, but it also would complicate the implementation. And if your test is very complicated and requires references to other tests for setup and cleanup by FMF ID, you can create tmt plan, which should already support these references.

**Comment by adiosnb** [comment]
The tmt developers agreed to implemented the feature in the last discussion. Adding some notes from the meeting:

- Motivation
  - Share setup and cleanup
  - Support scripts from external repositories as well?
    - Would need implement fetching remote repos
    - And also installing the possible requires
    - Recently brought by [Jiri Jaburek](mailto:jjaburek@redhat.com)
  - Better than having to manually set the test script
     - `test`;: ./setup.sh && ./test.sh && ./cleanup.sh
     - For local use case without external deps this already works, but it’s ugly :)
 - Benefit
   - Simple way to share setup/cleanup across multiple tests
   - Special handling of the phases? E.g. failing setup would result in error
   - Inheritance for many tests sharing the same setup and cleanup
 - Possibly related to DDT (Dependency Driven Testing)

**Comment by adiosnb** [comment]
@jscotka Considering various complications in remote linked tests, I propose splitting the solution into two PRs: #1966 with the local scripts and another with the remote tests.

What do you think about that and also about the implementation of #1966?

**Comment by seberm** [comment]
This issue and related PR #1966 was discussed on tmt hacking session today (17.12. 2024) , summary:

@jscotka  use case:
- Test setup is not relevant to the tmt plan but is closely related to individual tests
- Several test types, each test type needs a similar setup
- Possibly related plan of plans: Grouping a set of tests together
- Another approach: Using tmt tests as setup tasks
- Setup steps are duplicated if several tests with the same required-setup are executed one after the other
    - Some additional logic could/should/would have to be added to prevent this
    - Using separate tmt plans could cover the use case

This issue was added to the list of issues to be prioritized based on the user feedback (label `status | need prioritize`).

**Comment by happz** [comment]
My personal wishlist for this area, ideas I find naturally fitting the workflows of tmt:

* basically prepare/finish “steps”, just on test level
* allow using existing prepare/finish plugins in these “steps”. Do not implement shell/ansible/whatever test-level prepare/finish functionality: we already have it, find a way to collect phases on this level, and extract “run prepare phases” implementation from the prepare step.
* allow using tests in these “steps”
* I wanted tmt to support splitting plans into subplans grouping tests with same setup needed, but I’m no longer sure it’s needed or even desirable to support this. Why making the split? It’s an expensive prepare/finish task, or it’s destructive and would affect other tests. IMHO then it’s the responsibility of a user to do the split themselves by proper `discover` planning, tmt lacks context. But maybe there is a use case.
* “only once” type of prepare/finish task: if configured for many tests, it would run just once, before the first/after the last test. It sounds like a plan-level prepare/finish, but it’s attached to a test. If I pick just a single test, this task would run too; if I pick multiple tests, it would run as well, but only once.

**Comment by psss** [comment]
Copy & pasting a brainstorming idea from a duplicate issue #2132:

Another possible approach could be to introduce a new test check with custom commands. The config could look like this:

```yaml
check:
    how: custom
    before-script: ./do-something-before.sh
    after-script: ./do-something-after.sh
```

One possible advantage of this could be that the result aggregation (if desired for this feature) would be working out-of-the-box.

**Comment by psss** [comment]
The `prepare` and `finish` step approach would bring benefit of allowing fetching code from remote repos using the recently added `url`.

One more brainstorming idea: Use `tmt` test as a setup/cleanup task.

**Comment by sterlingalexander** [comment]
Just wanted to add an additional comment here, it looks like this is going towards a 'per test setup/cleanup' solution, and from a user perspective this is great.  Some tests require things that need to be set and subsequent tests probably should not inherit those settings.  

If something like this were implemented, it would decouple plan writers from test implementation details, and allow each test to 'just run' in an off-the-shelf manner.  In other words, if the test gets included in any arbitrary plan it will do it's own setup/cleanup and just run.  If plan-writers wanted to further adjust test behavior then they could learn what adjustment knobs/env vars are provided by the test.   Since most beaker users are accustom to writing setup/teardown code, maybe the implementation could be straighforward?

I also like the idea of test-dependencies, but maybe what is wanted here is more like test ordering guarantees?  If a test produces artifacts for subsequent tests to use, and assuming the subsequent tests know where to look for the artifacts, then if TMT were able to order them appropriately (or explicitly pull them in if they were missing and had a required 'run_after' or something similar) that would also help with the 'off the shelf' approach to running tests, further decoupling test plans from the actual tests.  

I don't like the idea of 'a specific test does setup for a batch of other tests so order the tests like this in the plan' because that just leaves strong coupling dependencies hidden for plan writers to trip over.  With all this in mind, maybe a way forward would be to allow for shared setup/teardown tasks to be run (or even inherited) across a fmf tree.  Then TMT could decide how to order tests that have the same setup/teardown requirements, with an ordering flag (run_{before,after} maybe) or some other mechanism to address the test-produced artifact case.  A toy example not using inheritance:

```
ls /kernel-tests/my-tests/
produce_thing_1/ produce_thing_2/ produce_thing_3/ test_protocol/ test_common/ test_regression/ verify_stuff/ 
```
```
ls /kernel-tests/my-tests/test_common/
common_tasks.fmf produce_things_tasks.fmf protocol_tasks.fmf
```

produce_thing_{1,2,3}

```
test: bash ./runtest.sh
  setup-require: /kernel-tests/my-tests/test_common/produce_things_tasks/setup
  finish-require: /kernel-tests/my-tests/test_common/produce_things_tasks/teardown
  run_before: /kernel-tests/my-tests/verify_stuff
  run_after:  # could be used to pull in 'required to run before this' dependencies?
```

produce_things_tasks.fmf

```
/setup:
  - All the stuff to do

/teardown:
  - All the stuff to do 
```

It's more work for the test writer, but imo it provides a much better experience for the test plan writers.  I would be willing to try and hack something together if people think this direction fits with TMTs overall design.    I also haven't fully considered the implications of yet another dependency chain, but there's no quicker way to learn something than to post a design and then write mediocre code to go with it, right?

**Comment by LecrisUT** [comment]
> I also like the idea of test-dependencies, but maybe what is wanted here is more like test ordering guarantees?

Yes, that will be part of the design, there was much more discussion in the surrounding parent issue https://github.com/teemtee/tmt/issues/4524, and the user interface will not be similar to the one here.

On that note, I hope to have an initial prototype in this week/next week and I'll post across all of these issues for more feedback then
